### PR TITLE
feat(display): assert when buffer is not aligned

### PR DESCRIPTION
**Pull request by XuNeo**
Help us review this PR! Anyone can [approve it or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### Description of the feature or fix

Default to 1 to make it behave same to lvgl v8.
To avoid confusions raised in lvgl/lv_binding_micropython#313

### Checkpoints
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html)

### CODE REVIEW

**Comment by kisvegabor** [comment]
Even SW render requires 4 byte alignment to read/write `uint32_t` data. Can we assume that it's fulfilled naturally?

**Comment by kisvegabor** [comment]
To approach is differently, if all memories are 4 byte aligned by default `LV_DRAW_BUF_ALIGN   4` shouldn't change anything. But at least it acts as a check if the address is really ok.

**Review by PGNetHun — APPROVED** [review]

**Comment by XuNeo** [comment]
For some CPU, the unaligned access is supported. I wonder if align to 4 will cause trouble like the one in micropython.

Set to 4 however will only cover the issue if unaligned access is not supported. The caller must be aware that it needs additional bytes in order to adjust alignment.

**Comment by kisvegabor** [comment]
> For some CPU, the unaligned access is supported. 

True, but probably way slower. So using `4` as default value would prevent hard to find performance issues too.

>  I wonder if align to 4 will cause trouble like the one in micropython.

To be honest I don't understand the MicroPython issue. Is it a problem if the allocated buffer is aligned?  With `LV_DRAW_BUF_ALIGN  4` nothing should happen (I assume `esp.heap_caps_malloc` returns at least word aligned memory)

> Set to 4 however will only cover the issue if unaligned access is not supported. The caller must be aware that it needs additional bytes in order to adjust alignment.

In `lv_display_set_*_buffers` I think an ASSERT without any extra magic if `LV_DRAW_BUF_ALIGN` is not fulfilled and `LV_DRAW_BUF_ALIGN  4` would be a good default value/behavior.

**Comment by PGNetHun** [comment]
Yes, hopefully `esp.heap_caps_malloc` function (by passing `MALLOC_CAP_DMA` flag) returns 32-bit aligned memory. 

Unfortunately `MALLOC_CAP_32BIT` flag is not present in `esp.MALLOC_CAP` enum in ESP-IDF 4.4.5:
image

Related document:
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/mem_alloc.html#dma-capable-memory

**Comment by kisvegabor** [comment]
If not 32 bit aligned is returned, I think it's better to assert. IMO expecting at least 4 byte align is quite reasonable.

**Comment by XuNeo** [comment]
> I think it's better to assert.

Do you mean we should assert in `lv_display_set_buffers` instead of secretly align it?

**Comment by kisvegabor** [comment]
> Do you mean we should assert in lv_display_set_buffers instead of secretly align it?

Yes, I think we can expect the user to do the alignment externally instead of doing hidden magic.

**Review by kisvegabor — DISMISSED** [review]

**Review by PGNetHun — APPROVED** [review]

File: src/display/lv_display.c [hunk 1]
```diff
@@ -398,11 +398,29 @@ void lv_display_set_draw_buffers(lv_display_t * disp, lv_draw_buf_t * buf1, lv_d
 void lv_display_set_buffers(lv_display_t * disp, void * buf1, void * buf2, uint32_t buf_size,
                             lv_display_render_mode_t render_mode)
 {
+    LV_ASSERT(buf1 != NULL);
+    if(buf1 == NULL) {
+        LV_LOG_ERROR("buf1 is NULL");
```
Comment (FASTSHIFT) [inline src/display/lv_display.c]: LV_ASSERT_NULL(buf1);

File: src/display/lv_display.c [hunk 2]
```diff
@@ -398,11 +398,29 @@ void lv_display_set_draw_buffers(lv_display_t * disp, lv_draw_buf_t * buf1, lv_d
...
     uint32_t w = lv_display_get_horizontal_resolution(disp);
     uint32_t h = lv_display_get_vertical_resolution(disp);
     LV_ASSERT(w != 0 && h != 0);
 
-    lv_color_format_t cf = lv_display_get_color_format(disp);
+    if(buf1 != lv_draw_buf_align(buf1, cf)) {
+        /* buf1 is not aligned according to LV_DRAW_BUF_ALIGN */
+        LV_LOG_ERROR("buf1 is not aligned");
+        return;
+    }
```
Comment (FASTSHIFT) [inline src/display/lv_display.c]: LV_ASSERT_MSG(buf1 == lv_draw_buf_align(buf1, cf), "buf1 is not aligned");

File: src/display/lv_display.c [hunk 3]
```diff
@@ -398,11 +398,29 @@ void lv_display_set_draw_buffers(lv_display_t * disp, lv_draw_buf_t * buf1, lv_d
...
+
+    if(buf2 && buf2 != lv_draw_buf_align(buf2, cf)) {
+        /* buf2 is not aligned according to LV_DRAW_BUF_ALIGN */
+        LV_LOG_ERROR("buf2 is not aligned");
+        return;
+    }
```
Comment (FASTSHIFT) [inline src/display/lv_display.c]: if(buf2 != NULL)
{
    LV_ASSERT_MSG(buf2 == lv_draw_buf_align(buf2, cf), "buf2 is not aligned");
}

Comment (XuNeo) [reply src/display/lv_display.c]: You have to add return when failed. Thus no need to use LV_ASSERT_MSG at all.

**Comment by XuNeo** [comment]
Added `LV_ASSERT_MSG` so it will assert if enabled.
CI needs https://github.com/lvgl/lvgl/pull/5392 to pass.

**Review by kisvegabor — APPROVED** [review]

**Review by FASTSHIFT — APPROVED** [review]

**Pull request merged**